src/artm/regularizer/topic_segmentation_ptdw.cc

```diff
@@ -16,107 +17,272 @@ namespace regularizer {
+  //double right_weights = 0.0;
+  //int l_topic, r_topic;  // topic ids on which maximum of the distribution is reached
+  //bool changes_topic = false;
+  //int main_topic = 0;
+  //auto main_topic_density = (*ptdw)(0, 0);
+  float tau = tau_;
+  float alpha = config_.merge_threshold();
+  bool merge_into_segments = config_.merge_into_segments();
+  std::list<int> &dot_positions = dot_positions_[item_index];
+  float average_len = 0.0f;
+  int seg_begin = 0;
+  for (auto it = dot_positions.begin(); it != dot_positions.end(); it++) {
+    average_len += *it - seg_begin;
+    seg_begin = *it + 1;
   }
-  for (int k = 0; k < topic_size; ++k) {
-    if (k == main_topic) {
-      (*ptdw)(0, k) = 1.0f;
-    } else {
-      (*ptdw)(0, k) = 0.0f;
-    }
+  average_len /= dot_positions.size();
+  //LOG(INFO) << dot_positions[dot_positions.size() - 1] << ' ' << local_token_size;
+  int c = 1;
+  if (inner_iter == 50) {
+    c = 1;
   }
-  for (int i = 0; i < h && i < local_token_size; ++i) {
-    for (int k = 0; k < topic_size; ++k) {
-      right_distribution[k] += copy_ptdw(i, k) * (1 - background_probability[i]);
+  for (int k = 0; k < c; k++){
+  int sen_begin = 0;
+  std::list<int>::iterator it = dot_positions.begin();
+  std::vector<float> prev_sen_subject(num_topics, 0.0f);
+  std::vector<float> distances(0, 0.0f);
+  bool handling_first_sen = true;
+  int prev_seg_len = 0;
+  while (sen_begin < local_token_size) {
+    int sen_end = *it;
+    int sen_len = sen_end - sen_begin;
+    if (sen_len <= 0) {
+      sen_begin = sen_end + 1;
+      it++;
+      continue;
     }
-    right_weights += 1 - background_probability[i];
-  }
-  for (int i = 1; i < local_token_size; ++i) {
-    for (int k = 0; k < topic_size; ++k) {
-      left_distribution[k] += copy_ptdw(i - 1, k) * (1 - background_probability[i - 1]);
-      right_distribution[k] -= copy_ptdw(i - 1, k) * (1 - background_probability[i - 1]);
+    std::vector<float> sen_subj(num_topics, 0.0f);
+    float norm_sum = 0.0f;
+    for (int i = sen_begin; i < sen_end; i++) {
+      float weight = 1.0f - background_probability[i];
+      //LOG(INFO) << weight << ' '<< background_probability[i];
+      //weight = 1.0f;
+      if (weight == 0.0f) continue;
+      norm_sum += weight * weight;
+      for (int t = 0; t < num_topics; t++) {
+        if (is_background_topic[t])
+          continue;
+        sen_subj[t] += copy_ptdw(i, t) * weight;
+      }
+    }
+    for (int t = 0; t < num_topics; t++) {
+      if (is_background_topic[t])
+        continue;
+      if (norm_sum != 0)
+        sen_subj[t] /= norm_sum;
     }
-    left_weights += 1 - background_probability[i - 1];
-    right_weights -= 1 - background_probability[i - 1];
-    if (i <= local_token_size - h) {
-      for (int k = 0; k < topic_size; ++k) {
-        right_distribution[k] += copy_ptdw(i + h - 1, k) * (1 - background_probability[i + h - 1]);
+    for (int i = sen_begin; i < sen_end; i++) {
+      float TMP = 0.0f;
```

Comment:
variables should be in lower case, try to use variables with name != temp/tmp and so on.